opnode/l1/downloader.go

```diff
@@ -0,0 +1,220 @@
+				return
+			}
+
+			txs := bl.Transactions()
+			bnr.Block = bl
+			bnr.Receipts = make([]*types.Receipt, len(txs))
+
+			for i, tx := range txs {
+				l1t.receiptTasks <- &receiptTask{BlockHash: id.Hash, TxHash: tx.Hash(), TxIndex: uint64(i), Dest: bnr}
+			}
+
+			// no receipts to fetch? Then we are done!
+			if len(txs) == 0 {
+				bnr.Finish(wrappedErr{nil})
+			}
+		}()
+	}
+	l1t.cacheLock.Unlock()
+
+	ch := make(chan wrappedErr)
+	sub := bnr.feed.Subscribe(ch)
+	select {
+	case wErr := <-ch:
+		if wErr.error != nil {
+			return nil, nil, wErr.error
+		}
+		return bnr.Block, bnr.Receipts, nil
+	case err := <-sub.Err():
+		return nil, nil, err
+	case <-ctx.Done():
+		return nil, nil, ctx.Err()
+	}
+}
+
+const maxReceiptRetry = 5
+
+func (l1t *downloader) newReceiptWorker() ethereum.Subscription {
+	return event.NewSubscription(func(quit <-chan struct{}) error {
+		for {
+			select {
+			case task := <-l1t.receiptTasks:
+				// scheduled tasks may be stale if other receipts of the block failed too many times
+				if task.Dest.finished > 0 { // no need for locks, a very rare stale download does not hurt
+					continue
+				}
+				// limit fetching to the task as a whole, and constrain to 10 seconds for receipt itself
+				ctx, cancel := context.WithTimeout(task.Dest.ctx, time.Second*10)
+				defer cancel()
+				receipt, err := l1t.chr.TransactionReceipt(ctx, task.TxHash)
+				if err != nil {
+					// if a single receipt fails out of the whole block, we can retry a few times.
+					if task.Retry >= maxReceiptRetry {
+						// Failed to get the receipt too many times, block fails!
+						task.Dest.Finish(wrappedErr{fmt.Errorf("failed to download receipt again, and reached max %d retries: %v", maxReceiptRetry, err)})
+						continue
+					} else {
+						task.Retry += 1
+						select {
+						case l1t.receiptTasks <- task:
+							// all good, retry scheduled successfully
+						default:
+							// failed to schedule, too much receipt work, stop block to relieve pressure.
+							task.Dest.Finish(wrappedErr{fmt.Errorf("receipt downloader too busy, not downloading receipt again (%d retries): %v", task.Retry, err)})
+							continue
+						}
+						continue
+					}
+				}
+				task.Dest.Receipts[task.TxIndex] = receipt
+				// We count the receipts we have so far (atomic, avoid parallel counting race condition)
+				total := atomic.AddUint32(&task.Dest.DownloadedReceipts, 1)
+				if total == uint32(len(task.Dest.Receipts)) {
+					// block completed without error!
+					task.Dest.Finish(wrappedErr{nil})
+					continue
+				}
+				// task completed, but block is not complete without other receipt tasks finishing first
+			case <-quit:
+				return nil
+			}
+		}
+	})
+}
+
+// AddReceiptWorkers can add or remove (negative value) worker routines to parallelize receipt downloads with.
+// It returns the number of active workers.
+func (l1t *downloader) AddReceiptWorkers(n int) int {
```

Comment:
Maybe good to add a `Close` that essentially does `AddReceiptWorkers(-len(l1t.receiptWorkers))` to clean up? Current closing would be a bit hacky: `l1t.AddReceiptWorkers(-l1t.AddReceiptWorkers(0))`.